src/dune_engine/scheduler.ml

```diff
@@ -643,8 +676,101 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Timer : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type sleeper
+
+  val await : sleeper -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> sleeper -> unit Fiber.t
+
+  val sleep : t -> float -> sleeper
+
+  val close : t -> unit
+end = struct
+  type sleeper = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable sleepers : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t sleeper =
```

Comment:
Don't we want to call `send_timer_completed` to make sure the count of active alarms is correct?
For that matter, why do we need that counter in the first place?

Reply:
> Don't we want to call send_timer_completed to make sure the count of active alarms is correct?

Good point.

> For that matter, why do we need that counter in the first place?

To prevent deadlocks I suppose

Reply:
I removed the pointless counter then.